doc/man/man1/total.1
.\" RCSid "$Id: total.1,v 1.12 2021/11/16 03:30:45 greg Exp $"
.TH TOTAL 1 2/3/95 RADIANCE
.SH NAME
total - sum up columns
.SH SYNOPSIS
.B total
[
.B \-m
][
.B \-sE
|
.B \-p
|
.B \-u
|
.B \-l
][
.B \-i{f|d}[N]
][
.B \-o{f|d}
][
.B "\-in M"
][
.B "\-on M"
][
.B \-tC
][
.B \-N
[
.B \-r
]]
[
file ..
]
.SH DESCRIPTION
.I Total
sums up columns of real numbers from one or more files
and prints out the result on its standard output.
.PP
By default,
.I total
computes the straigt sum of each input column, but multiplication
can be specified instead with the
.I \-p
option.
Likewise, the
.I \-u
option means find the upper limit (maximum), and
.I \-l
means find the lower limit (minimum).
.PP
Sums of powers can be computed by giving an exponent with the
.I \-s
option.
(Note that there is no space between the 
.I \-s
and the exponent.)
This exponent can be any real number, positive or negative.
The absolute value of the input is always taken before the
power is computed in order to avoid complex results.
Thus,
.I \-s1
will produce a sum of absolute values.
The default power (zero) is interpreted as a straight sum without
taking absolute values.
.PP
The
.I \-m
option can be used to compute the mean rather than the total.
For sums, the arithmetic mean is computed.
If a power is also specified using the
.I \-s
option, the inverse power will be applied to the averaged result.
For products, the geometric mean is computed.
(A logarithmic sum of absolute values is used to avoid overflow, and 
zero values are silently ignored.)
.PP
If the input data is binary, the
.I \-id
or
.I \-if
option may be given for 64-bit double or 32-bit float values, respectively.
Either option may be followed immediately by an optional
count, which defaults to 1, indicating the number of double or float
binary values to read per record on the input file.
(There can be no space between the option and this count.)\0
Similarly, the
.I \-od
and
.I \-of
options specify binary double or float output, respectively.
These options do not need a count, as this will be determined by the
number of input channels.
.PP
A count can be given as the number of lines to read before
computing a result.
Normally,
.I total
reads each file to its end before producing its result, 
but this behavior may be overridden by inserting blank lines in
the input.
For each blank input line, total produces a result as if the
end-of-file had been reached.
If two blank lines immediately follow each other, total closes
the file and proceeds to the next one (after reporting the result).
The
.I \-N
option (where N is a decimal integer) tells
.I total
to produce a result and reset the calculation after
every N input lines.
In addition, the
.I \-r
option can be specified to override reinitialization and thus
give a running total every N lines (or every blank line).
This option also turns off the usual output flushing at each total.
If the end of file is reached, the current total is printed
and the calculation is reset before the next file (with or without the
.I \-r
option).
.PP
The
.I \-in
option if present, will limit the number of input records read
(per input file).
The
.I \-on
option may be used to limit the total number of outut records produced.
.PP
The
.I \-tC
option can be used to specify the input and output tab character.
The default tab character is TAB.
.PP
If no files are given, the standard input is read.
.SH EXAMPLES
To compute the RMS value of colon-separated columns in a file:
.IP "" .2i
total \-t: \-m \-s2 input
.PP
To produce a running product of values from a file:
.IP "" .2i
total \-p \-1 \-r input
.SH BUGS
If the input files have varying numbers of columns, mean values
will certainly be off.
.I Total
will ignore missing column entries if the tab separator is a non-white
character, but cannot tell where a missing column should have been if
the tab character is white.
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
cnt(1), neaten(1), rcalc(1), rcollate(1), rlam(1), rsplit(1), tabfunc(1)


src/cal/total.c
#ifndef lint
static const char	RCSid[] = "$Id: total.c,v 1.15 2021/11/16 03:30:45 greg Exp $";
#endif
/*
 *  total.c - program to reduce columns of data.
 *
 *	5/18/88
 */

#include  <stdlib.h>
#include  <ctype.h>
#include  <math.h>
#include  "platform.h"
#include  "rtio.h"

#define  MAXCOL		8192		/* maximum number of columns */

#define  ADD		0		/* add numbers */
#define  MULT		1		/* multiply numbers */
#define  MAX		2		/* maximum */
#define  MIN		3		/* minimum */

double  init_val[] = {0., 1., -1e12, 1e12};	/* initial values */

long  incnt = 0;			/* limit number of input records? */
long  outcnt = 0;			/* limit number of output records? */

int  func = ADD;			/* default function */
double  power = 0.0;			/* power for sum */
int  mean = 0;				/* compute mean */
int  count = 0;				/* number to sum */
int  nbicols = 0;			/* number of binary input columns */
int  bocols = 0;			/* produce binary output columns */
int  tabc = '\t';			/* default separator */
int  subtotal = 0;			/* produce subtotals? */
int  nrecsout = 0;			/* number of records produced */

static int execute(char *fname);

int
main(
int  argc,
char  *argv[]
)
{
	int  status;
	int  a;

	for (a = 1; a < argc; a++)
		if (argv[a][0] == '-')
			if (argv[a][1] >= '0' && argv[a][1] <= '9')
				count = atoi(argv[a]+1);
			else
				switch (argv[a][1]) {
				case 'm':
					mean = !mean;
					break;
				case 'p':
					func = MULT;
					break;
				case 's':
					func = ADD;
					power = atof(argv[a]+2);
					break;
				case 'u':
					func = MAX;
					break;
				case 'l':
					func = MIN;
					break;
				case 't':
					tabc = argv[a][2];
					break;
				case 'r':
					subtotal = !subtotal;
					break;
				case 'i':
					switch (argv[a][2]) {
					case 'n':
						incnt = atol(argv[++a]);
						break;
					case 'a':
						nbicols = 0;
						break;
					case 'd':
						if (isdigit(argv[a][3]))
							nbicols = atoi(argv[a]+3);
						else
							nbicols = 1;
						if (nbicols > MAXCOL) {
							fprintf(stderr,
							"%s: too many input columns\n",
								argv[0]);
							exit(1);
						}
						break;
					case 'f':
						if (isdigit(argv[a][3]))
							nbicols = -atoi(argv[a]+3);
						else
							nbicols = -1;
						if (-nbicols > MAXCOL) {
							fprintf(stderr,
							"%s: too many input columns\n",
								argv[0]);
							exit(1);
						}
						break;
					default:
						goto userr;
					}
					break;
				case 'o':
					switch (argv[a][2]) {
					case 'n':
						outcnt = atol(argv[++a]);
						break;
					case 'a':
						bocols = 0;
						break;
					case 'd':
						bocols = 1;
						break;
					case 'f':
						bocols = -1;
						break;
					default:
						goto userr;
					}
					break;
				default:;
				userr:
					fprintf(stderr,
"Usage: %s [-m][-sE|p|u|l][-tC][-i{f|d}[N]][-o{f|d}][-count [-r]] [file..]\n",
							argv[0]);
					exit(1);
				}
		else
			break;
	if (mean) {
		if (func == MAX) {
			fprintf(stderr, "%s: average maximum?!\n", argv[0]);
			exit(1);
		}
		if (func == MIN) {
			fprintf(stderr, "%s: average minimum?!\n", argv[0]);
			exit(1);
		}
	}
	if (bocols)
		SET_FILE_BINARY(stdout);
	status = 0;
	if (a >= argc)
		status = execute(NULL) == -1 ? 1 : status;
	else
		for ( ; a < argc && (outcnt <= 0 || nrecsout < outcnt); a++)
			status = execute(argv[a]) == -1 ? 2 : status;
	exit(status);
}


static int
getrecord(			/* read next input record */
	double field[MAXCOL],
	FILE *fp
)
{
	char  buf[16*MAXCOL];
	char  *cp, *num;
	int   nf;
						/* reading binary input? */
	if (nbicols > 0)
		return(getbinary(field, sizeof(double), nbicols, fp));
	if (nbicols < 0) {
		float	*fbuf = (float *)buf;
		int	i;
		nf = getbinary(fbuf, sizeof(float), -nbicols, fp);
		for (i = nf; i-- > 0; )
			field[i] = fbuf[i];
		return(nf);
	}
						/* reading ASCII input */
	cp = fgets(buf, sizeof(buf), fp);
	if (cp == NULL)
		return(0);
	for (nf = 0; nf < MAXCOL; nf++) {
		while (isspace(*cp))
			cp++;
		if (!*cp)
			break;
		num = cp;
		while (*cp && !(isspace(tabc)?isspace(*cp):*cp==tabc))
			cp++;
		if (*cp)
			*cp++ = '\0';
		if (!*num || isspace(*num))
			field[nf] = init_val[func];
		else
			field[nf] = atof(num);
	}
	return(nf);
}


static void
putrecord(			/* write out results record */
	const double *field,
	int n,
	FILE *fp
)
{
						/* binary output? */
	if (bocols > 0) {
		putbinary(field, sizeof(double), n, fp);
		return;
	}
	if (bocols < 0) {
		float	fv;
		while (n-- > 0) {
			fv = *field++;
			putbinary(&fv, sizeof(float), 1, fp);
		}
		return;
	}
						/* ASCII output */
	while (n-- > 0) {
		fprintf(fp, "%.9g", *field++);
		if (n) fputc(tabc, fp);
	}
	fputc('\n', fp);
	if (!subtotal)
		fflush(fp);			/* flush unless -r */
}


static int
execute(			/* compute result */
char  *fname
)
{
	double	inpval[MAXCOL];
	double	tally[MAXCOL];
	short	rsign[MAXCOL];
	double  result[MAXCOL];
	int  n;
	int  nread, ncol;
	long  nlin, ltotal;
	FILE  *fp;
							/* open file */
	if (fname == NULL)
		fp = stdin;
	else if ((fp = fopen(fname, "r")) == NULL) {
		fprintf(stderr, "%s: cannot open\n", fname);
		return(-1);
	}
	if (nbicols)
		SET_FILE_BINARY(fp);
#ifdef getc_unlocked				/* avoid lock/unlock overhead */
	flockfile(fp);
#endif

	ltotal = 0;
	while (!feof(fp)) {
		if (ltotal == 0) {			/* initialize */
			if (func == MULT)	/* special case */
				for (n = 0; n < MAXCOL; n++) {
					tally[n] = 0.0;
					rsign[n] = 1;
				}
			else
				for (n = 0; n < MAXCOL; n++)
					tally[n] = init_val[func];
		}
		ncol = 0;
		for (nlin = 0; (count <= 0 || nlin < count) &&
				(incnt <= 0 || nlin < incnt) &&
				(nread = getrecord(inpval, fp)) > 0;
				nlin++) {
							/* compute */
			for (n = 0; n < nread; n++)
				switch (func) {
				case ADD:
					if (inpval[n] == 0.0)
						break;
					if (power != 0.0)
						tally[n] += pow(fabs(inpval[n]),power);
					else
						tally[n] += inpval[n];
					break;
				case MULT:
					if (inpval[n] == 0.0)
						rsign[n] = 0;
					else if (inpval[n] < 0.0) {
						rsign[n] = -rsign[n];
						inpval[n] = -inpval[n];
					}
					if (rsign[n])
						tally[n] += log(inpval[n]);
					break;
				case MAX:
					if (inpval[n] > tally[n])
						tally[n] = inpval[n];
					break;
				case MIN:
					if (inpval[n] < tally[n])
						tally[n] = inpval[n];
					break;
				}
			if (nread > ncol)
				ncol = nread;
		}
		if (nlin == 0)
			break;
						/* compute and print */
		ltotal += nlin;
		for (n = 0; n < ncol; n++) {
			result[n] = tally[n];
			if (mean) {
				result[n] /= (double)ltotal;
				if (func == ADD && power != 0.0 && result[n] != 0.0)
					result[n] = pow(result[n], 1.0/power);
			}
			if (func == MULT)
				result[n] = rsign[n] * exp(result[n]);
		}
		putrecord(result, ncol, stdout);
		++nrecsout;
		if (outcnt > 0 && nrecsout >= outcnt)
			break;
		if (!subtotal)
			ltotal = 0;
		if (incnt > 0 && nlin >= incnt)
			break;
	}
							/* close input */
	return(fclose(fp) == EOF ? 1 : 0);
}
